File: DemoObjC/DemoObjC/Test Cases/TestCase.m
//
//  TestCase.m
//  DemoObjC
//
// © 2026 Limelight Inc. All Rights Reserved.
//

#import "TestCase.h"
#import <LimelightSDK/LimelightSDK.h>
#import "BannerAdVC.h"
#import "InterstitialAdVC.h"
#import "RewardedAdVC.h"
#import "MAXBannerAdVC.h"
#import "MAXInterstitialAdVC.h"
#import "MAXRewardedAdVC.h"

// MARK: - Implementation

@implementation TestCase

- (instancetype)initWithTitle:(NSString *)title
                 storyboardId:(NSString *)storyboardId
         configurationClosure:(nullable void (^)(UIViewController *vc))configurationClosure {
    self = [super init];
    if (self) {
        _title = [title copy];
        _storyboardId = [storyboardId copy];
        _configurationClosure = [configurationClosure copy];
    }
    return self;
}

- (UIViewController *)makeViewController {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:[NSBundle mainBundle]];
    UIViewController *vc = [storyboard instantiateViewControllerWithIdentifier:self.storyboardId];
    vc.title = self.title;
    if (self.configurationClosure) {
        self.configurationClosure(vc);
    }
    return vc;
}

+ (NSDictionary<NSString *, NSArray<TestCase *> *> *)testCases {
    return @{
        @"In-App": [self allInAppCases],
        @"AppLovin MAX": [self allAppLovinCases],
    };
}

+ (nullable NSNumber *)indexOfSection:(NSString *)section {
    NSArray<NSString *> *keys = [self testCases].allKeys;
    NSUInteger index = [keys indexOfObject:section];
    return index != NSNotFound ? @(index) : nil;
}

+ (nullable NSString *)sectionKeyOfIndex:(NSInteger)index {
    NSArray<NSString *> *keys = [self testCases].allKeys;
    return (NSInteger)keys.count > index ? keys[index] : nil;
}

// MARK: - In-App Cases

+ (NSArray<TestCase *> *)allInAppCases {
    return [[self banners] arrayByAddingObjectsFromArray:
           [[self interstitials] arrayByAddingObjectsFromArray:
            [self rewarded]]];
}

// MARK: Banners

+ (NSArray<TestCase *> *)banners {
    return @[
        [[TestCase alloc] initWithTitle:@"Banner [320x50]"
                           storyboardId:@"BannerAdVC"
                   configurationClosure:^(UIViewController *vc) {
            BannerAdVC *bannerController = (BannerAdVC *)vc;
            bannerController.adSize = LLAdSize.banner_320x50;
            bannerController.adConfigId = @"123456789";
        }],
        [[TestCase alloc] initWithTitle:@"Banner [300x250]"
                           storyboardId:@"BannerAdVC"
                   configurationClosure:^(UIViewController *vc) {
            BannerAdVC *bannerController = (BannerAdVC *)vc;
            bannerController.adSize = LLAdSize.banner_300x250;
            bannerController.adConfigId = @"123456789";
        }],
    ];
}

// MARK: Interstitials

+ (NSArray<TestCase *> *)interstitials {
    return @[
        [[TestCase alloc] initWithTitle:@"Interstitial [Banner] [320x480]"
                           storyboardId:@"InterstitialAdVC"
                   configurationClosure:^(UIViewController *vc) {
            InterstitialAdVC *interstitialController = (InterstitialAdVC *)vc;
            interstitialController.adFormats = @[LLAdFormat.banner];
            interstitialController.adConfigId = @"123456789";
            interstitialController.adSize = CGSizeMake(320, 480);
        }],
        [[TestCase alloc] initWithTitle:@"Interstitial [Banner] [480x320]"
                           storyboardId:@"InterstitialAdVC"
                   configurationClosure:^(UIViewController *vc) {
            InterstitialAdVC *interstitialController = (InterstitialAdVC *)vc;
            interstitialController.adConfigId = @"123456789";
            interstitialController.adFormats = @[LLAdFormat.banner];
            interstitialController.adSize = CGSizeMake(480, 320);
        }],
        [[TestCase alloc] initWithTitle:@"Interstitial [Video] [480x320]"
                           storyboardId:@"InterstitialAdVC"
                   configurationClosure:^(UIViewController *vc) {
            InterstitialAdVC *interstitialController = (InterstitialAdVC *)vc;
            interstitialController.adConfigId = @"123456789";
            interstitialController.adFormats = @[LLAdFormat.video];
            interstitialController.adSize = CGSizeMake(480, 320);
        }],
        [[TestCase alloc] initWithTitle:@"Interstitial [Banner, Video] [Multi-Imp] [480x320]"
                           storyboardId:@"InterstitialAdVC"
                   configurationClosure:^(UIViewController *vc) {
            InterstitialAdVC *interstitialController = (InterstitialAdVC *)vc;
            interstitialController.adConfigId = @"123456789";
            interstitialController.adFormats = @[LLAdFormat.banner, LLAdFormat.video];
            interstitialController.adSize = CGSizeMake(480, 320);
        }],
    ];
}

// MARK: Rewarded

+ (NSArray<TestCase *> *)rewarded {
    return @[
        [[TestCase alloc] initWithTitle:@"Rewarded [Video] [480x320]"
                           storyboardId:@"RewardedAdVC"
                   configurationClosure:^(UIViewController *vc) {
            RewardedAdVC *rewardedAdController = (RewardedAdVC *)vc;
            rewardedAdController.adConfigId = @"123456789";
            rewardedAdController.adFormats = @[LLAdFormat.video];
            rewardedAdController.adSize = CGSizeMake(480, 320);
        }],
    ];
}

// MARK: - AppLovin MAX Cases

+ (NSArray<TestCase *> *)allAppLovinCases {
    return [[self alBanners] arrayByAddingObjectsFromArray:
           [[self alInterstitials] arrayByAddingObjectsFromArray:
            [self alRewarded]]];
}

// MARK: Banners

+ (NSArray<TestCase *> *)alBanners {
    return @[
        [[TestCase alloc] initWithTitle:@"Banner [320x50]"
                            storyboardId:@"MAXBannerAdVC"
                    configurationClosure:^(UIViewController *vc) {
            MAXBannerAdVC *bannerController = (MAXBannerAdVC *)vc;
            bannerController.adSize = CGSizeMake(320, 50);
            bannerController.adConfigId = @"123456789";
        }],
        [[TestCase alloc] initWithTitle:@"Banner [300x250]"
                            storyboardId:@"MAXBannerAdVC"
                    configurationClosure:^(UIViewController *vc) {
            MAXBannerAdVC *bannerController = (MAXBannerAdVC *)vc;
            bannerController.adSize = CGSizeMake(300, 250);
            bannerController.adConfigId = @"123456789";
        }],
    ];
}

// MARK: Interstitials

+ (NSArray<TestCase *> *)alInterstitials {
    return @[
        [[TestCase alloc] initWithTitle:@"Interstitial"
                            storyboardId:@"MAXInterstitialAdVC"
                    configurationClosure:^(UIViewController *vc) {
            MAXInterstitialAdVC *interstitialController = (MAXInterstitialAdVC *)vc;
            interstitialController.adConfigId = @"123456789";
        }],
    ];
}

// MARK: Rewarded

+ (NSArray<TestCase *> *)alRewarded {
    return @[
        [[TestCase alloc] initWithTitle:@"Rewarded"
                            storyboardId:@"MAXRewardedAdVC"
                    configurationClosure:^(UIViewController *vc) {
            MAXRewardedAdVC *rewardedAdController = (MAXRewardedAdVC *)vc;
            rewardedAdController.adConfigId = @"123456789";
        }],
    ];
}

@end

File: DemoSwift/DemoSwift/Test Cases/TestCase.swift
//
//  TestCase.swift
//  DemoSwift
//
// © 2026 Limelight Inc. All Rights Reserved.
//

import LimelightSDK
import UIKit

// MARK: - TestCase

struct TestCase {
    let title: String
    let storyboardId: String
    let configurationClosure: ((_ vc: UIViewController) -> Void)?

    func makeViewController() -> UIViewController {
        let vc = UIStoryboard(name: "Main", bundle: Bundle.main)
            .instantiateViewController(withIdentifier: storyboardId)
        vc.title = title
        configurationClosure?(vc)
        return vc
    }
}

extension TestCase {
    static let testCases: [String: [TestCase]] = [
        "In-App": allInAppCases,
        "AppLovin MAX": allAppLovinCases,
    ]

    static func index(ofSection: String) -> Int? {
        let keys = testCases.keys.map({ String($0) })
        return keys.firstIndex(of: ofSection)
    }

    static func sectionKey(ofIndex: Int) -> String? {
        let keys = testCases.keys.map({ String($0) })
        return keys.count > ofIndex ? keys[ofIndex] : nil
    }
}

// MARK: - In-App

extension TestCase {
    private static let allInAppCases: [TestCase] = banners + interstitials + rewarded

    // MARK: Banners
    private static let banners: [TestCase] = [
        TestCase(
            title: "Banner [320x50]",
            storyboardId: "BannerAdVC",
            configurationClosure: { vc in
                let bannerController = vc as! BannerAdVC
                bannerController.adSize = .banner_320x50
                bannerController.adConfigId = "123456789"
            }
        ),
        TestCase(
            title: "Banner [300x250]",
            storyboardId: "BannerAdVC",
            configurationClosure: { vc in
                let bannerController = vc as! BannerAdVC
                bannerController.adSize = .banner_300x250
                bannerController.adConfigId = "123456789"
            }
        ),
    ]

    // MARK: Interstitials
    private static let interstitials: [TestCase] = [
        TestCase(
            title: "Interstitial [Banner] [320x480]",
            storyboardId: "InterstitialAdVC",
            configurationClosure: { vc in
                let interstitialController = vc as! InterstitialAdVC
                interstitialController.adFormats = [.banner]
                interstitialController.adConfigId = "123456789"
                interstitialController.adSize = .init(width: 320, height: 480)
            }
        ),
        TestCase(
            title: "Interstitial [Banner] [480x320]",
            storyboardId: "InterstitialAdVC",
            configurationClosure: { vc in
                let interstitialController = vc as! InterstitialAdVC
                interstitialController.adConfigId = "123456789"
                interstitialController.adFormats = [.banner]
                interstitialController.adSize = .init(width: 480, height: 320)
            }
        ),
        TestCase(
            title: "Interstitial [Video] [480x320]",
            storyboardId: "InterstitialAdVC",
            configurationClosure: { vc in
                let interstitialController = vc as! InterstitialAdVC
                interstitialController.adConfigId = "123456789"
                interstitialController.adFormats = [.video]
                interstitialController.adSize = .init(width: 480, height: 320)
            }
        ),
        TestCase(
            title: "Interstitial [Banner, Video] [Multi-Imp] [480x320]",
            storyboardId: "InterstitialAdVC",
            configurationClosure: { vc in
                let interstitialController = vc as! InterstitialAdVC
                interstitialController.adConfigId = "123456789"
                interstitialController.adFormats = [.banner, .video]
                interstitialController.adSize = .init(width: 480, height: 320)
            }
        ),
    ]

    // MARK: Rewardeded
    private static let rewarded: [TestCase] = [
        TestCase(
            title: "Rewarded [Video] [480x320]",
            storyboardId: "RewardedAdVC",
            configurationClosure: { vc in
                let rewardedAdController = vc as! RewardedAdVC
                rewardedAdController.adConfigId = "123456789"
                rewardedAdController.adFormats = [.video]
                rewardedAdController.adSize = .init(width: 480, height: 320)
            }
        ),
    ]
}

// MARK: - AppLovin Max

extension TestCase {
    private static let allAppLovinCases: [TestCase] = alBanners + alInterstitials + alRewarded

    // MARK: Banners
    private static let alBanners: [TestCase] = [
        TestCase(
            title: "Banner [320x50]",
            storyboardId: "MAXBannerAdVC",
            configurationClosure: { vc in
                let bannerController = vc as! MAXBannerAdVC
                bannerController.adSize = CGSize(width: 320, height: 50)
                bannerController.adConfigId = "123456789"
            }
        ),
        TestCase(
            title: "Banner [300x250]",
            storyboardId: "MAXBannerAdVC",
            configurationClosure: { vc in
                let bannerController = vc as! MAXBannerAdVC
                bannerController.adSize = CGSize(width: 300, height: 250)
                bannerController.adConfigId = "123456789"
            }
        ),
    ]

    // MARK: Interstitials
    private static let alInterstitials: [TestCase] = [
        TestCase(
            title: "Interstitial",
            storyboardId: "MAXInterstitialAdVC",
            configurationClosure: { vc in
                let interstitialController = vc as! MAXInterstitialAdVC
                interstitialController.adConfigId = "123456789"
            }
        ),
    ]

    // MARK: Rewardeded
    private static let alRewarded: [TestCase] = [
        TestCase(
            title: "Rewarded",
            storyboardId: "MAXRewardedAdVC",
            configurationClosure: { vc in
                let rewardedAdController = vc as! MAXRewardedAdVC
                rewardedAdController.adConfigId = "123456789"
            }
        ),
    ]
}
